src/BlockTypes/ProductQuery.php

```diff
@@ -79,6 +79,60 @@ protected function initialize() {
 		);
 		add_filter( 'rest_product_query', array( $this, 'update_rest_query' ), 10, 2 );
 		add_filter( 'rest_product_collection_params', array( $this, 'extend_rest_query_allowed_params' ), 10, 1 );
+		add_filter( 'render_block_core/query', array( $this, 'add_navigation_id_directive' ), 10, 3 );
+		add_filter( 'render_block_core/query-pagination', array( $this, 'add_navigation_link_directives' ), 10, 3 );
+	}
+
+	/**
+	 * Mark the Product Query as an interactive region so it can be updated
+	 * during client-side navigation.
+	 *
+	 * @param string    $block_content The block content.
+	 * @param array     $block         The full block, including name and attributes.
+	 * @param \WP_Block $instance      The block instance.
+	 */
+	public function add_navigation_id_directive( $block_content, $block, $instance ) {
+		if ( self::is_woocommerce_variation( $block ) ) {
+			// Enqueue the Interactivity API runtime.
+			wp_enqueue_script( 'wc-interactivity' );
+
+			$p = new \WP_HTML_Tag_Processor( $block_content );
+
+			// Add `data-wc-navigation-id to the query block.
+			if ( $p->next_tag( array( 'class_name' => 'wp-block-query' ) ) ) {
+				$p->set_attribute( 'data-wc-interactive', true );
+				$p->set_attribute( 'data-wc-navigation-id', $block['attrs']['queryId'] );
```

Comment:
Maybe we should namespace this, to make it more future-proof, just in case. Something like:

```suggestion
				$p->set_attribute( 'data-wc-navigation-id', 'woo-products-' . $block['attrs']['queryId'] );
```

Reply:
Oh, you mean to avoid collisions between different blocks that could use the same ID for their `data-wc-navigation-id` attribute, right? That makes sense. 👍

Reply:
Fixed in https://github.com/woocommerce/woocommerce-blocks/pull/10200/commits/ade05add4196d7e506c55b433f9a74a8b66f79bd ✅